src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -274,47 +292,84 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
             }
             constantSourcesBuilder.Free();
 
-            // Run a syntax walk if any of the generators requested it
-            if (receiverCount > 0)
+
+            // PROTOTYPE(source-generators): we don't need to run at all if none of the inputs have changes.
+            var driverStateBuilder = new DriverStateTable.Builder(_state.StateTable, cancellationToken);
+
+            driverStateBuilder.SetInputValue(SharedInputNodes.Compilation, compilation);
+            driverStateBuilder.SetInputValue(SharedInputNodes.SyntaxTrees, compilation.SyntaxTrees);
+            driverStateBuilder.SetInputValue(SharedInputNodes.AnalyzerConfigOptions, _state.OptionsProvider);
+            driverStateBuilder.SetInputValue(SharedInputNodes.ParseOptions, _state.ParseOptions);
+            driverStateBuilder.SetInputValue(SharedInputNodes.AdditionalTexts, _state.AdditionalTexts);
+
+            // If we have syntax inputs, or receivers, bring them up to date
+            if (syntaxInputNodes.Count > 0 || receiverCount > 0)
             {
-                foreach (var tree in compilation.SyntaxTrees)
+                var builders = ArrayBuilder<ISyntaxTransformBuilder>.GetInstance(syntaxInputNodes.Count);
+                builders.AddRange(syntaxInputNodes.Select(n => n.GetBuilder(_state.StateTable)));
+
+                foreach ((var tree, var treeState) in driverStateBuilder.GetLatestStateTableForNode(SharedInputNodes.SyntaxTrees))
                 {
-                    var root = tree.GetRoot(cancellationToken);
+                    if (treeState == EntryState.Removed)
+                    {
+                        // we need to keep the removed tree entries, but can skip everything else
+                        foreach (var b in builders)
+                        {
+                            // PROTOTYPE(source-generators): we know we don't use the model in this case
+                            b.AddFilterFromPreviousTable(null!, EntryState.Removed);
```

Comment:
```suggestion
                            b.AddFilterFromPreviousTable(null, EntryState.Removed);
```

Can be removed if the other suggestions are taken in the `SyntaxTransformBuilder`